[nickczh] iP

royaliceblue opened the pull request.

royaliceblue committed:
Complete Level-0 (Rename, Greet, Exit)

Let's call the chatbot "Nick".

Let's implement an initial skeletal version of the ip that simply greets the user and exits.

Included *.class in the .gitignore file which will not be committed as
it should not be revision controlled.

royaliceblue committed:
Complete Level-1 (Echo)

Let's,

* Add a file containing ASCII content of name to the root of the
  project directory for uniqueness and customization.

* Improve skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

royaliceblue committed:
Complete Level-2 (Add, List)

Let's,
* Create a fixed size array ( String[100] ) to store the items.
* Store whatever text entered by the user and display them back to the user when requested.

royaliceblue committed:
Complete Level-3 (Mark as Done)

Let's,

* Add the ability to mark tasks as done.
* Add the ability to change the status back to not done.
* Implement extension (A-Classes) with Task class to represent tasks.

royaliceblue committed:
A-CodingStandard

Tweak the code to comply with a given coding standard.
Link: https://se-education.org/guides/conventions/java/basic.html

royaliceblue committed:
Nick.java: Change class name from Duke to Nick

Change class name from Duke to Nick for better customization.

royaliceblue committed:
Complete Level-4 (ToDos, Events, Deadlines)

Let's implement,
* ToDos: tasks without any date/time attached to it e.g., visit new theme park.
* Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm.
* Events: tasks that start at a specific date/time and ends at a specific date/time e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019
* A-Inheritance: use classes Todo, Deadline and Event classes to inherit from a Task class.

royaliceblue committed:
Complete A-TextUiTesting

Use the input/output redirection technique to semi-automate the testing of Duke.

Let's,
* Duplicate the name.txt file to the text-ui-test folder.
* Modify the while loop in Nick.java from while(true) to
  while(input.hasNextLine()).
* Add the input instructions to the input.txt file and expected results
  to the EXPECTED.TXT file in text-ui-test folder.

royaliceblue committed:
Complete A-CodeQuality

Critically examines the code and refactor to improve the code quality where necessary.

royaliceblue committed:
refactor: Add lines between functions

Kobot7 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+            switch (commandAction) {
+            case "bye":
+            case "list":
+            case "mark":
```
Perhaps place a "//Fallthrough" comment for case statements without a break statement?

royaliceblue replied:
Thanks for the catch!

Kobot7 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+            userTasks[Integer.parseInt(command.split(" ")[1]) - 1].markAsDone();
+            System.out.println("____________________________________________________________");
+            System.out.println("Nice! I've marked this task as done:");
+            System.out.println("\t" + "[" + userTasks[Integer.parseInt(command.split(" ")[1]) - 1].getStatusIcon() + "] " + userTasks[Integer.parseInt(command.split(" ")[1]) - 1].description);
```
Perhaps keep the line length to less than 110-120 characters?

Kobot7 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+
+        boolean hasQuit = false;
+        String command, commandAction;
+        Task[] userTasks = new Task[100];
```
Consider making this a constant?

Kobot7 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+            printByeMsg();
+            return false;
+        case ("list"):
+            System.out.println("____________________________________________________________");
```
Consider making this a constant?

classskipper351 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+        Scanner input = new Scanner(System.in);
+
+        while (input.hasNextLine()) {
+            command = input.nextLine();
+            commandAction = command.split(" ")[0];
```
maybe CommandAction is better in coding standards

classskipper351 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+
+        switch (taskType) {
+        case "todo":
+            taskName = command.substring(taskDescriptionIndex);
+            task = new Todo(taskName);
+            userTasks[taskCount] = task;
+            break;
+        case "deadline":
+            int deadlineIndex = command.indexOf("/by") + DEADLINE_OFFSET_IDX;
+            taskName = command.substring(taskDescriptionIndex, taskDescriptionEndIndex);
+            String deadline = command.substring(deadlineIndex);
+            task = new Deadline(taskName, deadline);
+            userTasks[taskCount] = task;
+            break;
+        case "event":
+            int fromIndex = command.indexOf("/from");
+            int toIndex = command.indexOf("/to");
+            taskName = command.substring(taskDescriptionIndex, taskDescriptionEndIndex);
+            String from = command.substring(fromIndex + FROM_OFFSET_IDX, toIndex - 1);
+            String to = command.substring(toIndex + TO_OFFSET_IDX);
+            task = new Event(taskName, from, to);
+            userTasks[taskCount] = task;
```
I would  consider extracting codes in cases to a single method

royaliceblue replied:
Good recommendation! Thank you.

classskipper351 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+            case "list":
+            case "mark":
+            case "unmark":
+                boolean continueProgram = continueExecution(command, taskCount, userTasks);
+                if (continueProgram) {
+                    continue;
```
Maybe name boolean values with the prefix of "is"

classskipper351 commented on src/main/java/Nick.java:
```diff
@@ -0,0 +1,142 @@
+                if (continueProgram) {
+                    continue;
+                }
+                hasQuit = true;
+            }
+            if (hasQuit) {
+                break;
+            }
+
+            taskExecution(command, commandAction, userTasks);
```
a little confused by the control flow

classskipper351 reviewed (commented):
pretty good encapsulation and inheritance and processing of strings in total

royaliceblue committed:
refactor: Add UserInterface class

Let's,
* Abstract printByeMsg(), printAddTaskMsg(), printIntroName(),
  printIntroMsg() to the UserInterface class.

royaliceblue committed:
refactor: Implement PR review improvement comments

royaliceblue committed:
Complete Level-5 (Handle Errors)

Let's,

* Teach chatbot to deal with errors such as incorrect inputs entered by
  the user.

* Handle invalid commands by throwing NickException() object and
  handling StringIndexOutOfBoundsException

* Create a new NickException class

royaliceblue committed:
Merge branch 'branch-Level-5'

royaliceblue committed:
Complete A-Packages (Organize into Packages)

Let's,
* Create a command package containing our Deadline, Event, Task and Todo
  classes.

royaliceblue committed:
Merge branch 'branch-A-Packages'

royaliceblue committed:
Complete Level 6 (Delete) with A-Collections

Let's,
* Use the Java Collections class ArrayList<Task> to store our tasks.

royaliceblue committed:
Complete Level 7 (Save)

royaliceblue committed:
Merge branch 'branch-Level-6'

royaliceblue committed:
Merge branch 'branch-Level-7'

Let's,
* Fix merge conflict.

royaliceblue committed:
Complete A-MoreOOP (Use More OOP)

Let's,

* Create Ui class to deal with interactions with the user.

* Create Storage class to deal with loading tasks from the file and
  saving tasks in the file.

* Create Parser class to deal with making sense of the user command.

* Create TaskList class to contain the task list eg: it has operations
  to add/delete tasks in the list.

* Implement stretch goal to extract *Command classes which inherit from
  an abstract Command class and store them in command package.

royaliceblue committed:
refactor: Package classes

Let's,

* Package ui, storage and parser class

royaliceblue committed:
Complete Level-8 (Dates and Times)

Let's,

* Store deadline dates as a java.time.LocalDate (or java.time.LocalDateTime) in your task objects.
* Accept dates in a format such as yyyy-mm-dd format (e.g., 2019-10-15) and print in a different format such as MMM dd yyyy e.g., (Oct 15 2019).

royaliceblue committed:
Complete Level-9 (Find)

Let's,

* Give users a way to find a task by searching for a keyword.

royaliceblue committed:
Complete A-JavaDoc (JavaDoc)

Let's,

* Add JavaDoc comments to the code.

royaliceblue committed:
Merge pull request #1 from nickczh/branch-Level-8

Complete Level-8 (Dates and Times)

royaliceblue committed:
Merge branch 'master' into branch-Level-9

royaliceblue committed:
Merge branch 'master' into branch-A-JavaDoc

royaliceblue committed:
Merge pull request #2 from nickczh/branch-Level-9

Complete branch-Level-9 (Find)

royaliceblue committed:
Merge branch 'master' into branch-A-JavaDoc

royaliceblue committed:
Merge pull request #3 from nickczh/branch-A-JavaDoc

Complete branch-A-JavaDoc (JavaDoc)

royaliceblue committed:
Complete A-UserGuide (User Guide)

Let's,

* Add a User Guide to the project under docs/README.md

royaliceblue committed:
refactor: Update README.md

royaliceblue committed:
refactor: Update README.md

okkhoy commented on src/main/java/Nick/Nick.java:
```diff
@@ -0,0 +1,53 @@
+package Nick;
```
should be in lower case

royaliceblue replied:
Thanks for the catch Prof Akshay! Have made the corresponding improvements to my code.

okkhoy commented on src/main/java/Nick/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        int DEADLINE_OFFSET_IDX = 4;
+        int taskDescriptionEndIndex = arguments.indexOf("/") - 1;
+        int deadlineIndex = arguments.indexOf("/by") + DEADLINE_OFFSET_IDX;
+        taskName = arguments.substring(0, taskDescriptionEndIndex);
```
consider SLAPping out some of the parsing stuff.

okkhoy commented on src/main/java/Nick/command/FindCommand.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        System.out.println("Here are the matching tasks in your list:");
+        tasks.tasks.stream()
+                .filter((t) -> t.description.contains(findDescription))
+                .forEach(result -> System.out.println((counter++) + ". " + result.toString()));
```
😎

okkhoy commented on src/main/java/Nick/command/ListCommand.java:
```diff
@@ -0,0 +1,29 @@
+public class ListCommand extends Command {
+
+    /**
+     * Iterate through each task object in the ArrayList and prints out each task object.
```
`Iterate**s** ...`

okkhoy commented on src/main/java/Nick/storage/Storage.java:
```diff
@@ -0,0 +1,145 @@
+            String to = command.substring(toIndex + TO_OFFSET_IDX);
+            task = new Event(taskName, from, to, taskDone);
+            userTasks.add(task);
+            //taskCount++;
```
remove dead code

okkhoy commented on src/main/java/Nick/storage/Storage.java:
```diff
@@ -0,0 +1,145 @@
+                    fw.write("deadline " + "| " + done + " | " + userTasks.get(i).description + " /by " + ((Deadline) userTasks.get(i)).getDeadline() + "\n");
+                }
+                else if (userTasks.get(i) instanceof Event) {
+                    fw.write("event " + "| " + done + " | " + userTasks.get(i).description + " /from " + ((Event) userTasks.get(i)).getFrom() + " /to " + ((Event) userTasks.get(i)).getTo() + "\n");
```
long line. try to "fold" it

okkhoy reviewed (commented):
Generally neet code (👍) except for a few minor misses that I have pointed out. Pls see if you can fix it before the submission.

royaliceblue committed:
refactor: Clean up code for final version

royaliceblue committed:
refactor: Remove name.txt from Ui class

royaliceblue committed:
refactor: Clean up code for exception

royaliceblue committed:
refactor: Add more exceptions to commands

Let's,

* Include NullPointerException and StringIndexOutOfBoundsException to
  commands

royaliceblue committed:
docs: Update README.md
